Feature/improved upgrade notes (#115)
wireapp/wire-docs: src/how-to/upgrade/config-references/wire-server-5.24.0.md

Deploy it **before** the `wire-server` upgrade. If `wire-server` is upgraded first with `gundeck.config.redis.host` already pointing at `redis-ephemeral`, gundeck won't be able to connect.

```bash
d helm upgrade --install redis-ephemeral ./charts/redis-ephemeral \
  --values ./values/redis-ephemeral/prod-values.example.yaml
```

Defaults are fine. Nothing should be carried over from `databases-ephemeral`, it's a different chart.

The old `databases-ephemeral` chart isn't auto-removed. Don't uninstall it yet either, wait until everything else is verified. See "Cleanup" at the bottom.

### 2. Edit `brig.config.rabbitmq` in the wire-server values

In `values/wire-server/values.yaml`:

```yaml
brig:
  config:
    rabbitmq:
      host: rabbitmq-host-or-service
```

In `values/wire-server/secrets.yaml`:

```yaml
brig:
  secrets:
    rabbitmq:
      username: wire-server
      password: <rabbitmq-password>
```

`port` defaults to `5672`, only set it if the RabbitMQ instance listens somewhere else.

> Why: starting at `5.23`, `rabbitmq` is deployed as an external service instead of in-cluster. `brig` can't rely on the in-cluster default service name anymore, so the hostname has to be set explicitly.

Reviewer: ```suggestion
Why: for those using our ansible infrastructure package, starting at `5.23`, `rabbitmq` is deployed as an external service instead of in-cluster. `brig` can't rely on the in-cluster default service name anymore, so the hostname has to be set explicitly.
```

### 3. Add the new `background-worker` config

`background-worker` needs a few new fields. In `values/wire-server/values.yaml`:

```yaml
background-worker:
  config:
    federationDomain: example.com  # match the federation domain used everywhere else (brig, etc.)
    cassandraBrig:
      host: <cassandra-host-or-service>  # same value as the existing cassandra.host
    cassandraGalley:
      host: <cassandra-host-or-service>  # same value as the existing cassandra.host
    postgresql:
      host: <postgresql-host-or-service>
```

In `values/wire-server/secrets.yaml`:

```yaml
background-worker:
  secrets:
    pgPassword: <postgresql-password>
```

To grab the existing PostgreSQL password from a typical deploy:

```bash
d kubectl get secret postgresql-external -o jsonpath='{.data.password}' | base64 -d
```

Why: `background-worker` now runs jobs that need PostgreSQL access and that talk directly to the Cassandra keyspaces of `brig` and `galley`. The federation domain is needed for federation-related background tasks.

> > **Warning about the `postgresMigration.conversation` default.** At this release the `background-worker` chart defaults `postgresMigration.conversation` to `postgresql`. That default must **not** be left in place when conversations haven't been migrated yet. The data is still in Cassandra, so a `postgresql` setting points the worker at an empty table. See the migration section below. This is the bug that may have caused conversations to disappear at `5.25` for installs that didn't migrate.

Reviewer: ```suggestion
> **Warning about the `postgresMigration.conversation` default.** At this release the `background-worker` chart defaults `postgresMigration.conversation` to `postgresql`. That default must **not** be left in place when conversations haven't been migrated yet. If they have not yet been migrated, the data will still be in Cassandra, so a `postgresql` setting points the worker at an empty table. See the migration section below. This is the bug that may have caused conversations to disappear at `5.25` for installs that didn't migrate.
```

### 4. Update `gundeck.config.redis.host`

> The Redis service hostname changed from `{{ .Release.Name }}-master` to `{{ .Release.Name }}`. With the standard `redis-ephemeral` release name, the in-cluster service is just `redis-ephemeral` now (used to be `databases-ephemeral-redis-ephemeral-master`).

Reviewer: ```suggestion
For those using the in-cluster Redis service shipped with our bundle, The Redis service hostname has changed from `{{ .Release.Name }}-master` to `{{ .Release.Name }}`. With the standard `redis-ephemeral` release name, the in-cluster service is just `redis-ephemeral` now (used to be `databases-ephemeral-redis-ephemeral-master`).
```

background-worker:
  config:
    migrateConversations: false
    postgresMigration:
      conversation: migration-to-postgresql
```

Then run:

```bash
d helm upgrade --install wire-server ./charts/wire-server --timeout=15m0s \
  --values ./values/wire-server/values.yaml \
  --values ./values/wire-server/secrets.yaml
```

Once it's set to `migration-to-postgresql`, do not switch back to `cassandra`. New conversations from this point on are written to PostgreSQL, reads still come from Cassandra.

### Step 2: run the actual migration

In `values/wire-server/values.yaml`:

```yaml
background-worker:
  config:
    migrateConversations: true
    postgresMigration:
      conversation: migration-to-postgresql
```

Then run:

```bash
d helm upgrade --install wire-server ./charts/wire-server --timeout=15m0s \
  --values ./values/wire-server/values.yaml \
  --values ./values/wire-server/secrets.yaml
```

The `background-worker` pods restart and start moving data. This can take a long time on a database with a lot of conversations.

Watch the logs (look for `finished migration`):

```bash
d kubectl logs deployment/background-worker --tail=2000 | grep migrate-conversations
```

> Or watch the metrics, both of these should hit `1.0`:

Reviewer: add a 'if you need to pull metrics manually' link here.

* `wire_local_convs_migration_finished`
* `wire_user_remote_convs_migration_finished`

### Step 3: switch reads over to PostgreSQL

Once the metrics are at `1.0`, in `values/wire-server/values.yaml`:

```yaml
galley:
  config:
    postgresMigration:
      conversation: postgresql
background-worker:
  config:
    migrateConversations: false
    postgresMigration:
      conversation: postgresql
```

Then run:

```bash
d helm upgrade --install wire-server ./charts/wire-server --timeout=15m0s \
  --values ./values/wire-server/values.yaml \
  --values ./values/wire-server/secrets.yaml
```

> From now on reads and writes both go to PostgreSQL. This configuration must be kept on every subsequent upgrade.

Reviewer: ```suggestion
From this point forward reads and writes both go to PostgreSQL. This configuration must be kept on every subsequent upgrade.
```

> From now on reads and writes both go to PostgreSQL. This configuration must be kept on every subsequent upgrade.

Reviewer: the last part of this sentence is unclear.

## Optional changes

> ### `background-worker` tunables

Reviewer: ```suggestion
### `background-worker` PostGreSQL tunables
```

```

`gundeck` pods should all be `Running`:

```bash
d kubectl get pods | grep gundeck
```

If the conversation migration is already done, the row count in PostgreSQL should be non-zero. SSH into one of the postgres nodes:

```bash
ssh <postgres-node> "sudo -u postgres psql -d wire-server -c 'SELECT COUNT(*) FROM conversation;'"
```

And before calling it done, log in on the webapp and on a mobile client, send a message in an existing conversation, confirm conversations are still there. The whole point of the migration is to keep that data, so it must be verified for real.

## Cleanup

Once `redis-ephemeral` looks healthy, the old `databases-ephemeral` chart can go. Nothing uninstalls it automatically.

First confirm nothing still references it:

```bash
d helm list -A | grep databases-ephemeral
d kubectl get svc | grep databases-ephemeral
grep -r "databases-ephemeral" values/wire-server/values.yaml
```

The grep on `values/wire-server/values.yaml` should come back empty. If it doesn't, fix the override first, then come back.

Then:

```bash
d helm uninstall databases-ephemeral
```

## Disk space note

> Each upgrade in this series re-runs `setup-offline-sources`, which copies the new release's binaries, container images, and debs into `/opt/assets` on the assethost. After a few versions, the assethost runs out of space and the playbook fails with `no space left on device`.

Reviewer: ```suggestion
For those of you using our ansible based deployment package, each upgrade in this series re-runs `setup-offline-sources`, which copies the new release's binaries, container images, and debs into `/opt/assets` on the assethost. After a few versions, the assethost runs out of space and the playbook fails with `no space left on device`.
```
